src/pmdas/farm/pmda.c

```diff
@@ -0,0 +1,1907 @@
+/*
+ * Seagate Field Accessible Reliability Metrics (FARM) Log
+ *
+ * Copyright (c) 2023 Red Hat.
+ * 
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+ 
+ #include "pmapi.h"
+ #include "pmda.h"
+ #include "domain.h"
+ 
+ #include "pmdafarm.h"
+ 
+ static int _isDSO = 1; /* for local contexts */
+ static char *farm_setup_lsblk;
+ static char *farm_setup_smartctl;
+ 
+ pmdaIndom indomtable[] = {
+ 	{ .it_indom = DISK_INDOM },
+ };
+ 
+ #define INDOM(x) (indomtable[x].it_indom)
+ 
+ /*
+ * All metrics supported by this PMDA - one table entry for each metric
+ */
+pmdaMetric metrictable[] = {
+	/* ATA - PAGE 0: FARM log header */
+	{ .m_desc = {
+		PMDA_PMID(CLUSTER_ATA_LOG_HEADER, LOG_VERSION),
+		PM_TYPE_STRING, DISK_INDOM, PM_SEM_DISCRETE,
+		PMDA_PMUNITS(0,0,0,0,0,0) }, },
+	{ .m_desc = {
+		PMDA_PMID(CLUSTER_ATA_LOG_HEADER, PAGES_SUPPORTED),
+		PM_TYPE_U64, DISK_INDOM, PM_SEM_DISCRETE,
+		PMDA_PMUNITS(0,0,0,0,0,0) }, },
+	{ .m_desc = {
+		PMDA_PMID(CLUSTER_ATA_LOG_HEADER, LOG_SIZE),
```

Comment:
Once the instance domains are tackled, this metric table will shrink significantly.
Next issue to tackle then (easier with fewer metrics) is the units and semantics need to be reviewed and update - I'm seeing lots of things that are bytes, times, etc here.  These need correct units and the semantics are either PM_SEM_COUNTER or PM_SEM_INSTANT for most metrics here I think.